### PR TITLE
VERSION: release runc 1.0.0

### PR DESCRIPTION
**Pull request by cyphar**
**This PR needs 5 votes to pass:**

- [ ] Michael Crosby <michael@docker.com> (@crosbymichael)
- [x] Mrunal Patel <mpatel@redhat.com> (@mrunalp)
- [ ] Daniel, Dao Quang Minh <dqminh89@gmail.com> (@dqminh)
- [x] Qiang Huang <h.huangqiang@huawei.com> (@hqhq)
- [x] Aleksa Sarai <cyphar@cyphar.com> (@cyphar)
- [x] Akihiro Suda <akihiro.suda.cz@hco.ntt.co.jp> (@AkihiroSuda)
- [x] Kir Kolyshkin <kolyshkin@gmail.com> (@kolyshkin)

```
v1.0.0 -- "A wizard is never late, nor is he early, he arrives precisely when he means to."

This release fixes a few bugs found (almost all related to cgroupv2
handling), and is the first non-rc release of runc in 5 years
(v1.0.0-rc1 was released in 2016). It's been a very long road, and we
thank the many contributors and maintainers that helped us get to this
point (approximately 422 people in total).

As runc follows Semantic Versioning, we will endeavor to not make any
breaking changes without bumping the major version number of runc.

However, it should be noted that Go API usage of runc's internal
implementation (libcontainer) is *not* covered by this policy -- for
historical reasons, this code was not moved into an "internal" package
(this feature did not exist in Go at the time) and because certain
projects currently depend on this, we have not yet moved this code into
an internal package. Despite this, we reserve the right to make breaking
changes in our Go APIs (though we will note such changes in our
changelog, and will try to avoid needless disruption if possible).

Breaking changes:

 * Removed libcontainer/configs.Device* identifiers (deprecated since rc94,
   use libcontainer/devices) (#2999)
 * Removed libcontainer/system.RunningInUserNS function (deprecated since
   rc94, use libcontainer/userns) (#2999)

Deprecations:

 * The usage of relative paths for mountpoints will now produce a warning
   (such configurations are outside of the spec, and in future runc will
   produce an error when given such configurations). (#2917, #3004)

Bugfixes:

 * cgroupv2: devices: rework the filter generation to produce consistent
   results with cgroupv1, and always clobber any existing eBPF
   program(s) to fix `runc update` and avoid leaking eBPF programs
   (resulting in errors when managing containers).  (#2951)
 * cgroupv2: correctly convert "number of IOs" statistics in a
   cgroupv1-compatible way. (#2965, #2967, #2968, #2964)
 * cgroupv2: support larger than 32-bit IO statistics on 32-bit architectures.
 * cgroupv2: wait for freeze to finish before returning from the freezing
   code, optimize the method for checking whether a cgroup is frozen. (#2955)
 * cgroups/systemd: fixed "retry on dbus disconnect" logic introduced in rc94
 * cgroups/systemd: fixed returning "unit already exists" error from a systemd
   cgroup manager (regression in rc94) (#2997, #2996)

Improvements:

 * cgroupv2: support SkipDevices with systemd driver (#2958, #3019)
 * cgroup/systemd: return, not ignore, stop unit error from Destroy (#2946)
 * Fix all golangci-lint failures. (#2781, #2962)
 * Make "runc --version" output sane even when built with `go get` or
   otherwise outside of our build scripts. (#2962)
 * cgroups: set SkipDevices during runc update (so we don't modify
   cgroups at all during `runc update`). (#2994)
 * cgroup1: blkio: support BFQ weights. (#3010)
 * cgroupv2: set per-device io weights if BFQ IO scheduler is available.
   (#3022)

Thanks to the following people who made this release possible:

 * Adrian Reber <areber@redhat.com>
 * Akihiro Suda <akihiro.suda.cz@hco.ntt.co.jp>
 * Aleksa Sarai <cyphar@cyphar.com>
 * Antti Kervinen <antti.kervinen@intel.com>
 * Daniel, Dao Quang Minh <dqminh89@gmail.com>
 * Enrico Weigelt <info@metux.net>
 * Kir Kolyshkin <kolyshkin@gmail.com>
 * Michael Crosby <michael@thepasture.io>
 * Mrunal Patel <mrunal@me.com>
 * Peter Hunt <pehunt@redhat.com>
 * Qiang Huang <h.huangqiang@huawei.com>
 * Sebastiaan van Stijn <github@gone.nl>
 * Shiming Zhang <wzshiming@foxmail.com>
 * Yashpal Choudhary <yashpal.c1995@gmail.com>

Signed-off-by: Aleksa Sarai <cyphar@cyphar.com>
```

Signed-off-by: Aleksa Sarai <cyphar@cyphar.com>

### CODE REVIEW

**Review by dqminh — DISMISSED** [review]

**Comment by AkihiroSuda** [comment]
Triggering CIs
https://github.com/moby/moby/pull/42450 https://github.com/containerd/containerd/pull/5552 https://github.com/containerd/nerdctl/pull/231 https://github.com/moby/buildkit/pull/2143

**Comment by AkihiroSuda** [comment]
Can we have an official announcement on https://opencontainers.org/posts/announcements/ ? 🎉

File: VERSION [hunk 1]
```diff
@@ -1 +1 @@
-1.0.0-rc95+dev
+1.0.0+dev
```
Comment (vrothberg) [inline VERSION:1]: Is the `+dev` suffix intentional?

Comment (cyphar) [reply VERSION:1]: There are two commits -- the first is the 1.0.0 release then one to switch back to development so builds of subsequent commits will show "+dev" rather than "1.0.0". This is what our release PRs usually look like.

**Comment by h-vetinari** [comment]
> 
> 
> Can we have an official announcement on https://opencontainers.org/posts/announcements/ ? 🎉

I think that would be a great idea, _and_ deserve a longer victory lap than the proposed release notes in the OP. There's a lot of history across:
* 7+ years (5 in rc-mode) since https://github.com/moby/moby/pull/4452
* 2 repo changes
* ~95~ 15 release candidates
* various CVEs
* the hook drama
* cgroups v2, rootless, etc. etc.

and concluding this epic journey would IMO really deserve some celebratory commemoration.

File: VERSION (hunk 1 above)
Comment (vrothberg) [reply VERSION:1]: Looks like I am ready for the lunch break. Thanks :)

**Comment by cyphar** [comment]
> 95 15 release candidates

Please stop rubbing it in. :sweat_smile: :sweat_smile: :sweat_smile: 

I promise I've actually read the SemVer spec now.

**Comment by dims** [comment]
@cyphar is it worth updating the changelog to include https://github.com/opencontainers/runc/pull/2965 and https://github.com/opencontainers/runc/pull/2968 ?

**Comment by cyphar** [comment]
@dims It does mention them, though I linked to the PRs not the issues:

>  * cgroupv2: correctly convert "number of IOs" statistics in a
>   cgroupv1-compatible way. (#2967, #2964)

**Comment by dims** [comment]
> @dims It does mention them, though I linked to the PRs not the issues:
> 
> > * cgroupv2: correctly convert "number of IOs" statistics in a
> >   cgroupv1-compatible way. (#2967, #2964)

whoops! thanks.

**Review by AkihiroSuda — CHANGES_REQUESTED** [review]
Moby CI is failing 😭

https://github.com/moby/moby/pull/42450#issuecomment-852094051

> === RUN TestBuildMultiStageCopy/copy_to_newdir
--- FAIL: TestBuildMultiStageCopy/copy_to_newdir (0.42s)
build_test.go:182: {"stream":"Step 1/8 : FROM busybox AS base"}
{"stream":"\n"}
{"stream":" ---\u003e 1c35c4412082\n"}
{"stream":"Step 2/8 : RUN mkdir existingdir"}
{"stream":"\n"}
{"stream":" ---\u003e Running in a65f6cb4628e\n"}
{"stream":"Removing intermediate container a65f6cb4628e\n"}
{"errorDetail":{"message":"failed to create shim: OCI runtime create failed: container_linux.go:380: starting container process caused: process_linux.go:545: container init caused: process_linux.go:508: setting cgroup config for procHooks process caused: failed to call BPF_PROG_ATTACH (BPF_CGROUP_DEVICE, BPF_F_ALLOW_MULTI): can't attach program: invalid argument: unknown"},"error":"failed to create shim: OCI runtime create failed: container_linux.go:380: starting container process caused: process_linux.go:545: container init caused: process_linux.go:508: setting cgroup config for procHooks process caused: failed to call BPF_PROG_ATTACH (BPF_CGROUP_DEVICE, BPF_F_ALLOW_MULTI): can't attach program: invalid argument: unknown"}

**Comment by AkihiroSuda** [comment]
I guess we might be hitting some kernel limit value related to cgroup2 ebpf?

**Comment by cyphar** [comment]
Hmmm, possibly -- or there's something wrong with how cgroup limits are interacting with hooks -- I'll take a look at this. I guess we need to postpone 1.0.0 :cry:.

**Comment by cyphar** [comment]
Ah, I think it's a kernel version thing. I forgot to add `BPF_F_REPLACE` support detection in #2951. I'll work up a PR tomorrow.

**Review by crosbymichael — DISMISSED** [review]

**Comment by kolyshkin** [comment]
>As runc follows Semantic Versioning, we will endevour to not make any

endeavour/endeavor?

**Review by mrunalp — DISMISSED** [review]

**Comment by kolyshkin** [comment]
> * cgroupv2: various SetFreezer optimisations. (#2955)

This is actually a bug fix, too (we never waited for the freeze to finish, now we do). So I'd reword it as

```
* cgroupv2: wait for freeze to finish, optimize (#2955)
```

(or something along the lines)

**Comment by kolyshkin** [comment]
I suggest adding https://github.com/opencontainers/runc/pull/2946 to "Improvements" section, like this:
```
* cgroup/systemd: return, not ignore, stop unit error from Destroy (#2946)
```

**Comment by cyphar** [comment]
@kolyshkin Yeah I saw those PRs but there wasn't a per-written changelog entry so I wasn't quite sure what was the right description. I'll update the draft release description.

**Review by kolyshkin — DISMISSED** [review]
LGTM

**Comment by kolyshkin** [comment]
I have added the impact/changelog label to https://github.com/opencontainers/runc/pull/2999, and provided the changelog entry in there, to be added to 1.0.0 release notes.

**Review by Hotdogszbg — APPROVED** [review]

**Comment by thaJeztah** [comment]
I'd like to have some eyes on https://github.com/opencontainers/runc/pull/3004 for consideration, as I'm seeing more reports of that validation change breaking users (for various reasons). I generally "agree" with the validation, but wonder if it's worth breaking users over for being "correct".

**Comment by thaJeztah** [comment]
Quick blurb before I forget; in LinuxKit, they found a change in behaviour w.r.t. permissions (I think in rc94/rc95). It's possible this was intentional (I haven't found time to bisect); in case someone knows if this was an intentional change, or a regression; https://github.com/linuxkit/linuxkit/pull/3684#issuecomment-855057520 /cc @rn @djs55

**Comment by cyphar** [comment]
@thaJeztah I looked into it (see #3025 and my comments in the original issue). It seems to be a result of us removing `/dev/console` from the default allow list a few releases back, which was an intentional change. They can work around it fairly easily so there's nothing for us to do on this end.

**Comment by thaJeztah** [comment]
Thanks for looking, @cyphar. I thought "better safe than sorry" 🤗

**Comment by kolyshkin** [comment]
The only thing left is https://github.com/opencontainers/runc/pull/3022, I think it needs to be merged now as it fixes an issue with (and provides a test case for) recently merged https://github.com/opencontainers/runc/pull/3010#issuecomment-860302432

**Comment by cyphar** [comment]
@opencontainers/runc-maintainers I've rebased this on top of the all the necessary fixes we've merged in the past few weeks, so we need to re-do the vote.

**Comment by hqhq** [comment]
LGTM

**Review by hqhq — APPROVED** [review]

**Comment by cyphar** [comment]
LGTM.

**Review by AkihiroSuda — APPROVED** [review]
LGTM

**Review by kolyshkin — APPROVED** [review]
LGTM

**Review by mrunalp — APPROVED** [review]

**Comment by mrunalp** [comment]
LGTM

**Review by thaJeztah — APPROVED** [review]
LGTM (nb)

**Comment by cyphar** [comment]
That's 5. I'll do the release on Monday.

**Review by tao12345666333 — APPROVED** [review]
LGTM (nb)

**Comment by kolyshkin** [comment]
I took a liberty to remove

> * cgroupv2: various SetFreezer optimisations. (#2955)

line from the release notes, as it's already described above (i.e. a duplicate).

**Comment by h-vetinari** [comment]
This could still use an announcement on https://opencontainers.org/, the mailing list, etc... Not sure that's worth an issue, but can open one of course.

**Comment by cyphar** [comment]
While the opencontainers website is deployed from https://github.com/opencontainers/opencontainers.org, I'll ping @caniszczyk since usually someone else drafts the blog announcements.
